### PR TITLE
Catch git fetching failures, fix checkout of tag/SHA1

### PR DESCRIPTION
**Pull request by LDong-Arm**
Fixes: #38

This PR:
* adds git failure catching to `check_and_clone_repo()`
* fixes checkout of tag/SHA1 (uncovered by the failure catching)

### CODE REVIEW

File: psa_builder.py [hunk 1]
```diff
@@ -204,16 +204,35 @@ def check_and_clone_repo(name, deps, dir):
...
         logging.info("Cloning %s repo", name)
         cmd = ["git", "-C", dir, "clone", "-b", gitref, deps.get(name)[0]]
-        run_cmd_and_return(cmd)
-        logging.info("Cloned %s repo successfully", name)
+        ret = run_cmd_and_return(cmd)
+        if ret == 0:
+            logging.info("Cloned %s repo successfully", name)
+        else:
+            logging.critical("Failed to clone %s repo, error: %d", name, ret)
+            sys.exit(1)
```
Comment (jainvikas8) [inline psa_builder.py:212]:
``` suggestion
        if ret != 0:
            logging.critical("Failed to clone %s repo, error: %d", name, ret)
            sys.exit(1)

         logging.info("Cloned %s repo successfully", name)

File: psa_builder.py [hunk 2]
```diff
@@ -204,16 +204,35 @@ def check_and_clone_repo(name, deps, dir):
...
         cmd = ["git", "-C", join(dir, name), "fetch"]
-        run_cmd_and_return(cmd)
-        logging.info("%s repo exists, checking out %s...", name, gitref)
+        ret = run_cmd_and_return(cmd)
+        if ret != 0:
+            logging.critical(
+                "Failed to fetch the latest %s, error: %d", name, ret
+            )
+            sys.exit(1)
+        logging.info("Checking out %s...", gitref)
```
Comment (jainvikas8) [inline psa_builder.py:222]:
``` suggestion
        if ret != 0:
            logging.critical(
                "Failed to fetch the latest %s, error: %d", name, ret
            )
            sys.exit(1)

        logging.info("Checking out %s...", gitref)

File: psa_builder.py [hunk 3]
```diff
@@ -204,16 +204,35 @@ def check_and_clone_repo(name, deps, dir):
...
+        if ret != 0:
+            # gitref might be a tag or SHA1 which we checkout directly
+            cmd = ["git", "-C", join(dir, name), "checkout", gitref]
+            ret = run_cmd_and_return(cmd)
+            if ret != 0:
+                logging.critical(
+                    "Failed to checkout %s, error: %d", gitref, ret
+                )
+                sys.exit(1)
         logging.info("Checked out %s successfully", gitref)
```
Comment (jainvikas8) [inline psa_builder.py:236]:
``` suggestion
        ret = run_cmd_and_return(cmd)
        if ret != 0:
            logging.critical(
                    "Failed to checkout %s, error: %d", head, ret
                )

            # gitref might be a tag or SHA1 which we checkout directly
            cmd = ["git", "-C", join(dir, name), "checkout", gitref]
            ret = run_cmd_and_return(cmd)
            if ret != 0:
                logging.critical(
                    "Failed to checkout %s, error: %d", gitref, ret
                )
                sys.exit(1)
            
            logging.info("Checked out %s successfully", gitref)
        else:
            logging.info("Checked out %s successfully", head)

**Review by jainvikas8 — CHANGES_REQUESTED** [review]

**Comment by jainvikas8** [comment]
We can squash these commits together and have one description

File: psa_builder.py (hunk 3 above)
Comment (LDong-Arm) [reply psa_builder.py:236]: Not sure about logging an error at the first `if ret != 0` - users may be confused and think there's a real problem if they are not familiar. And in my personal opinion, printing `gitref` as the message is only for a rough reference, but I'm fine with `head` (i.e. `origin/<gitref>`) too.

Comment (LDong-Arm) [reply psa_builder.py]: The "repo exists, checking out " above (even before this PR) only prints `gitref` without "origin/", so it's fine to do the same here.

File: psa_builder.py (hunk 1 above)
Comment (LDong-Arm) [reply psa_builder.py]: Done

File: psa_builder.py (hunk 2 above)
Comment (LDong-Arm) [reply psa_builder.py:223]: Done. Added an empty line after the `if` block.

**Review by LDong-Arm — COMMENTED** [review]
@jainvikas8 Thanks for the review. I've just made changes.

**Review by jainvikas8 — APPROVED** [review]

**Pull request merged**